File: 04-estructuras-datos/10-heaps-y-colas-prioridad/scala/ejemplos/ColaPrioridad.scala
import scala.collection.mutable.PriorityQueue

object ColaPrioridad:
  case class Incidencia(codigo: String, prioridad: Int)

  def main(args: Array[String]): Unit =
    given Ordering[Incidencia] = Ordering.by(i => -i.prioridad)
    val cola = PriorityQueue.empty[Incidencia]
    // El Ordering define qué incidencia debe salir primero.
    cola.enqueue(Incidencia("INC-003", 3), Incidencia("INC-001", 1))
    println(cola.dequeue())

File: 04-estructuras-datos/10-heaps-y-colas-prioridad/java/ejemplos/ColaPrioridad.java
import java.util.Comparator;
import java.util.PriorityQueue;

public class ColaPrioridad {
    record Incidencia(String codigo, int prioridad) {}

    public static void main(String[] args) {
        PriorityQueue<Incidencia> cola = new PriorityQueue<>(Comparator.comparingInt(Incidencia::prioridad));

        // PriorityQueue entrega primero el elemento que cumple el criterio de mayor prioridad.
        cola.offer(new Incidencia("INC-003", 3));
        cola.offer(new Incidencia("INC-001", 1));
        cola.offer(new Incidencia("INC-002", 2));

        System.out.println("Atender primero: " + cola.poll());
    }
}
